### PR TITLE
Add warn message while met inaccessible metrics

### PR DESCRIPTION
**Pull request by Colstuwjx**
As [#159](https://github.com/prometheus/snmp_exporter/issues/159#issuecomment-301778779)  said, I create this PR to make `snmp generator` throws warning message while met inaccessible metrics.

### CODE REVIEW

**Comment by Colstuwjx** [comment]
@SuperQ , @brian-brazil   would you mind take a review?

**Comment by SuperQ** [comment]
I would take this a step farther, and automatically ignore any metric marked `ACCESS_READONLY` or `ACCESS_READWRITE`.

**Comment by RichiH** [comment]
Please carry a counter for inaccessible oids.

Richard

Sent by mobile; excuse my brevity and the wall of text Gmail appends by
default.

File: generator/tree.go [hunk 1]
```diff
@@ -145,6 +145,12 @@ func generateConfigModule(cfg *ModuleConfig, node *Node, nameToNode map[string]*
 			if !ok {
 				return // Unsupported type.
 			}
+
+			// Unaccessible metrics should throw warning
+			if n.Access == "ACCESS_NOACCESS" || n.Access == "unknown" {
```
Comment (SuperQ) [inline generator/tree.go]: Maybe make a new function `metricAccess()` that is similar to `metricType` that deals with access data.

**Comment by SuperQ** [comment]
@RichiH This is on the generator side, no metrics needed, this will filter them out before giving them to the exporter.

**Comment by Colstuwjx** [comment]
it seems that skipping the unaccessible metrics would not hurt the `index`\`lookup`, the `snmp generator` ONLY skip the unaccessible things.

@SuperQ  @brian-brazil  Please let me know if there is any problems, thanks.

File: generator/tree.go [hunk 2]
```diff
@@ -107,6 +107,16 @@ func metricType(t string) (string, bool) {
 	}
 }
 
+func metricAccess(a string) (string, bool) {
+	switch a {
+	case "ACCESS_READONLY", "ACCESS_READWRITE":
```
Comment (brian-brazil) [inline generator/tree.go]: Also create per RFC 3578 section 7.3

File: generator/tree.go [hunk 3]
```diff
@@ -107,6 +107,16 @@ func metricType(t string) (string, bool) {
 	}
 }
 
+func metricAccess(a string) (string, bool) {
```
Comment (brian-brazil) [inline generator/tree.go]: This should only return a bool

File: generator/tree.go [hunk 4]
```diff
@@ -145,6 +155,12 @@ func generateConfigModule(cfg *ModuleConfig, node *Node, nameToNode map[string]*
 			if !ok {
 				return // Unsupported type.
 			}
+
+			_, ok = metricAccess(n.Access)
+			if !ok {
+				return // Unaccessible metrics.
```
Comment (brian-brazil) [inline generator/tree.go]: Inaccessible

File: generator/tree.go (hunk 2 above)
Comment (Colstuwjx) [inline generator/tree.go]: Yes, according to the [RFC 2578](https://tools.ietf.org/html/rfc2578#section-7.3), `read-create` is the superset of `red-write`, many thanks.

**Comment by Colstuwjx** [comment]
OK, I'd like to add a test case in `tree_test.go`.

File: generator/tree.go [hunk 5]
```diff
@@ -145,6 +156,12 @@ func generateConfigModule(cfg *ModuleConfig, node *Node, nameToNode map[string]*
 			if !ok {
 				return // Unsupported type.
 			}
+
+			ok = metricAccess(n.Access)
```
Comment (brian-brazil) [inline generator/tree.go]: You can put this in the if

File: generator/tree.go [hunk 6]
```diff
@@ -145,6 +156,11 @@ func generateConfigModule(cfg *ModuleConfig, node *Node, nameToNode map[string]*
 			if !ok {
 				return // Unsupported type.
 			}
+
+			if ok = metricAccess(n.Access); !ok {
```
Comment (SuperQ) [inline generator/tree.go]: I think this can be `if metricAccess(n.Access) {`

File: generator/tree.go [hunk 7]
```diff
@@ -157,8 +157,7 @@ func generateConfigModule(cfg *ModuleConfig, node *Node, nameToNode map[string]*
 				return // Unsupported type.
 			}
 
-			ok = metricAccess(n.Access)
-			if !ok {
+			if ok = metricAccess(n.Access); !ok {
```
Comment (brian-brazil) [inline generator/tree.go]: `if metric(n.Access) {`

File: generator/tree_test.go [hunk 8]
```diff
@@ -262,6 +262,55 @@ func TestGenerateConfigModule(t *testing.T) {
...
+		// Tables with accessible & inaccessible.
+		{
+			node: &Node{Oid: "1", Label: "root",
+				Children: []*Node{
+					{Oid: "1.1", Label: "table",
+						Children: []*Node{
+							{Oid: "1.1.1", Label: "tableEntry",
+								Children: []*Node{
+									{Oid: "1.1.1.1", Access: "ACCESS_NOACCESS", Label: "tableNoAccessIndex", Type: "INTEGER"},
+									{Oid: "1.1.1.2", Access: "ACCESS_CREATE", Label: "tableCreateDescr", Type: "OCTETSTR"},
```
Comment (brian-brazil) [inline generator/tree_test.go]: I'd make these all of type integer to keep focused on what we're actually testing here.

**Comment by Colstuwjx** [comment]
I introduced a test case for accessible/inaccessible metrics, and `go test -run TestGenerateConfigModule` output is:

```
...

	tree_test.go:572: GenerateConfigModule: difference in case 4
	tree_test.go:574: Got: walk:
		- "1"
		metrics:
		- name: tableCreate
		  oid: 1.1.1.2
		  type: gauge
		- name: tableReadOnly
		  oid: 1.1.1.4
		  type: gauge
		- name: tableReadWrite
		  oid: 1.1.1.5
		  type: gauge
	tree_test.go:576: Wanted: walk:
		- "1"
		metrics:
		- name: tableCreate
		  oid: 1.1.1.2
		  type: gauge
		- name: tableWriteOnly
		  oid: 1.1.1.3
		  type: gauge
		- name: tableReadOnly
		  oid: 1.1.1.4
		  type: counter
		- name: tableReadWrite
		  oid: 1.1.1.5
		  type: gauge
		- name: tableNotify
		  oid: 1.1.1.6
		  type: gauge

...

FAIL
exit status 1
FAIL	github.com/prometheus/snmp_exporter/generator	0.006s
```

Is that okay?

File: generator/tree_test.go [hunk 9]
```diff
@@ -284,27 +284,27 @@ func TestGenerateConfigModule(t *testing.T) {
...
 					{
-						Name: "tableCreateDescr",
+						Name: "tableCreate",
 						Oid:  "1.1.1.2",
-						Type: "OctetString",
+						Type: "gauge",
 					},
 					{
-						Name: "tableWriteOnlyId",
+						Name: "tableWriteOnly",
```
Comment (brian-brazil) [inline generator/tree_test.go]: Should this and Notify be missing from the outputted metrics?

**Comment by SuperQ** [comment]
Please also squash these commits.

**Comment by brian-brazil** [comment]
Github allows commits to be squashed these days, so unless there's vendoring involved where we want to keep some commits there's no need to ask the contributor to squash.

File: generator/tree_test.go (hunk 9 above)
Comment (Colstuwjx) [inline generator/tree_test.go]: I might get the answer. Due to our introduced `access` mechanism, our test cases also generate its own access code, and things doesn't work anymore...

File: generator/tree_test.go [hunk 10]
```diff
@@ -262,6 +262,55 @@ func TestGenerateConfigModule(t *testing.T) {
...
+			out: &config.Module{
+				Walk: []string{"1"},
+				Metrics: []*config.Metric{
+					{
+						Name: "tableCreate",
+						Oid:  "1.1.1.2",
+						Type: "gauge",
+					},
+					{
+						Name: "tableWriteOnly",
```
Comment (SuperQ) [inline generator/tree_test.go]: I think this one should not be in the output.

File: generator/tree_test.go [hunk 11]
```diff
@@ -262,6 +262,55 @@ func TestGenerateConfigModule(t *testing.T) {
...
+						Oid:  "1.1.1.4",
+						Type: "gauge",
+					},
+					{
+						Name: "tableReadWrite",
+						Oid:  "1.1.1.5",
+						Type: "gauge",
+					},
+					{
+						Name: "tableNotify",
```
Comment (SuperQ) [inline generator/tree_test.go]: Same here, this should be dropped from the output.

File: generator/tree.go [hunk 12]
```diff
@@ -145,6 +155,11 @@ func generateConfigModule(cfg *ModuleConfig, node *Node, nameToNode map[string]*
 			if !ok {
 				return // Unsupported type.
 			}
+
+			if !metricAccess(n.Access) {
+				return // Inaccessible metrics.
```
Comment (Colstuwjx) [inline generator/tree.go]: After adding some debug log here , the whole test case output is:

```
WARN[0000] Error, root with  is inaccessible             source=tree.go:160
WARN[0000] Error, root with  is inaccessible             source=tree.go:160
WARN[0000] Error, root with  is inaccessible             source=tree.go:160
WARN[0000] Error, OCTETSTR with  is inaccessible         source=tree.go:160
WARN[0000] Error, INTEGER with  is inaccessible          source=tree.go:160
WARN[0000] Error, NETADDR with  is inaccessible          source=tree.go:160
WARN[0000] Error, IPADDR with  is inaccessible           source=tree.go:160
WARN[0000] Error, COUNTER with  is inaccessible          source=tree.go:160
WARN[0000] Error, GAUGE with  is inaccessible            source=tree.go:160
WARN[0000] Error, TIMETICKS with  is inaccessible        source=tree.go:160
WARN[0000] Error, COUNTER64 with  is inaccessible        source=tree.go:160
WARN[0000] Error, BITSTRING with  is inaccessible        source=tree.go:160
WARN[0000] Error, UINTEGER with  is inaccessible         source=tree.go:160
WARN[0000] Error, UNSIGNED32 with  is inaccessible       source=tree.go:160
WARN[0000] Error, INTEGER32 with  is inaccessible        source=tree.go:160
WARN[0000] Error, MacAddress with  is inaccessible       source=tree.go:160
WARN[0000] Error, tableNoAccess with ACCESS_NOACCESS is inaccessible  source=tree.go:160
WARN[0000] Error, tableWriteOnly with ACCESS_WRITEONLY is inaccessible  source=tree.go:160
WARN[0000] Error, tableNotify with ACCESS_NOTIFY is inaccessible  source=tree.go:160
WARN[0000] Error, tableIndex with  is inaccessible       source=tree.go:160
WARN[0000] Error, tableFoo with  is inaccessible         source=tree.go:160
WARN[0000] Error, octetIndex with  is inaccessible       source=tree.go:160
WARN[0000] Error, octetFoo with  is inaccessible         source=tree.go:160
WARN[0000] Error, bitstringIndex with  is inaccessible   source=tree.go:160
WARN[0000] Error, bitstringFoo with  is inaccessible     source=tree.go:160
WARN[0000] Error, ipaddrIndex with  is inaccessible      source=tree.go:160
WARN[0000] Error, ipaddrFoo with  is inaccessible        source=tree.go:160
WARN[0000] Error, netaddrIndex with  is inaccessible     source=tree.go:160
WARN[0000] Error, netaddrFoo with  is inaccessible       source=tree.go:160
WARN[0000] Error, physaddress48Index with  is inaccessible  source=tree.go:160
WARN[0000] Error, physaddress48Foo with  is inaccessible  source=tree.go:160
WARN[0000] Error, octetFoo with  is inaccessible         source=tree.go:160
```

I'm still wondering the result...

Comment (brian-brazil) [inline generator/tree.go]: I guess you'll need to add access to all of the other test cases.

Comment (Colstuwjx) [inline generator/tree.go]: Yes, happy to see things work well again. BTW, why will the `tableWriteOnly` \\ `tableNotify` be skipped? I'm still not clear about that.

**Comment by brian-brazil** [comment]
Thanks!

**Comment by SuperQ** [comment]
Woot, this is great, thanks for the PR!

**Pull request merged**